From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Wed, 3 Apr 2019 17:07:24 +0200
Subject: [PATCH] media: coda: fill up bitstream ring buffer on decoder stop

When V4L2_CMD_DEC_STOP is issued, copy pending output buffers into the
bitstream ring buffer before setting the stream end flag.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/media/platform/coda/coda-common.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/media/platform/coda/coda-common.c b/drivers/media/platform/coda/coda-common.c
index 23f3a13c018a..ecd65530df2a 100644
--- a/drivers/media/platform/coda/coda-common.c
+++ b/drivers/media/platform/coda/coda-common.c
@@ -1381,6 +1381,19 @@ static int coda_decoder_cmd(struct file *file, void *fh,
 			/* Mark last buffer */
 			buf->flags |= V4L2_BUF_FLAG_LAST;
 
+			/*
+			 * TODO: Copy *all* currently queued output
+			 * buffers into the bitstream ring buffer, not
+			 * just enough to allow a PIC_RUN to succeed.
+			 * If not all buffers fit, the marked last
+			 * buffer will stop decoding later.
+			 */
+			if (ctx->use_bit) {
+				mutex_lock(&ctx->bitstream_mutex);
+				coda_fill_bitstream(ctx, NULL);
+				mutex_unlock(&ctx->bitstream_mutex);
+			}
+
 			if (v4l2_m2m_num_src_bufs_ready(ctx->fh.m2m_ctx) == 0) {
 				coda_dbg(1, ctx, "all remaining buffers queued\n");
 				stream_end = true;
